torchgeo/datasets/geo.py

```diff
@@ -990,35 +953,33 @@ def __init__(
         .. versionadded:: 0.4
             The *transforms* parameter.
         """
-        super().__init__(transforms)
         self.datasets = [dataset1, dataset2]
         self.collate_fn = collate_fn
+        self.transforms = transforms
 
         for ds in self.datasets:
             if not isinstance(ds, GeoDataset):
                 raise ValueError('IntersectionDataset only supports GeoDatasets')
 
-        self.crs = dataset1.crs
-        self.res = dataset1.res
-
-        # Merge dataset indices into a single index
-        self._merge_dataset_indices()
-
-    def _merge_dataset_indices(self) -> None:
-        """Create a new R-tree out of the individual indices from two datasets."""
-        i = 0
-        ds1, ds2 = self.datasets
-        for hit1 in ds1.index.intersection(ds1.index.bounds, objects=True):
-            for hit2 in ds2.index.intersection(hit1.bounds, objects=True):
-                box1 = BoundingBox(*hit1.bounds)
-                box2 = BoundingBox(*hit2.bounds)
-                box3 = box1 & box2
-                # Skip 0 area overlap (unless 0 area dataset)
-                if box3.area > 0 or box1.area == 0 or box2.area == 0:
-                    self.index.insert(i, tuple(box3))
-                    i += 1
-
-        if i == 0:
+        dataset2.crs = dataset1.crs
+        dataset2.res = dataset1.res
+
+        # Spatial intersection
+        index1 = dataset1.index.reset_index()
+        index2 = dataset2.index.reset_index()
+        self.index = gpd.overlay(index1, index2, how='intersection')
+
+        # Temporal intersection
+        datetime_1 = pd.IntervalIndex(self.index.pop('datetime_1'))
+        datetime_2 = pd.IntervalIndex(self.index.pop('datetime_2'))
+        mint = np.maximum(datetime_1.left, datetime_2.left)
+        maxt = np.minimum(datetime_1.right, datetime_2.right)
+        self.index.index = pd.IntervalIndex.from_arrays(
+            mint, maxt, closed='both', name='datetime'
+        )
+        # self.index = self.index[maxt > mint]
```

Comment:
Adding this line of code causes a couple tests to fail (point datasets, box edge overlap). But presumably this line of code is needed in cases where some files have spatial overlap but not temporal overlap? I don't think our tests yet exercise this, may need to add a new test.